pep-9999.rst

```diff
@@ -0,0 +1,393 @@
+       AllTheDefaults[int, complex, str, int] ==
+       AllTheDefaults[int, complex, str, int, bool]
+   )  # All valid
+
+This cannot be enforce at runtime for functions, for now, but in the
+future, this might be possible (see `Interaction with PEP
+695 <#interaction-with-pep-695>`__).
+
+``Generic`` ``TypeAlias``\ es
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``Generic`` ``TypeAlias``\ es should be able to be further subscripted
+following normal subscription rules. If a ``TypeVar`` has a default
+that hasn't been overridden it should be treated like it was
+substituted into the ``TypeAlias``. However, it can be specialised
+further down the line.
+
+.. code:: py
+
+   class SomethingWithNoDefaults(Generic[T, T2]): ...
+
+   MyAlias: TypeAlias = SomethingWithNoDefaults[int, DefaultStrT]  # valid
+   reveal_type(MyAlias)        # type is SomethingWithNoDefaults[int, str]
+   reveal_type(MyAlias[bool])  # type is SomethingWithNoDefaults[int, bool]
+
+   MyAlias[bool, int]  # Invalid: too many arguments passed to MyAlias
+
+Subclassing
+~~~~~~~~~~~
+
+Subclasses of ``Generic``\ s with ``TypeVar``\ s that have defaults
+behave similarly to ``Generic`` ``TypeAlias``\ es.
+
+.. code:: py
+
+   class SubclassMe(Generic[T, DefaultStrT]): ...
+
+   class Bar(SubclassMe[int, DefaultStrT]): ...
+   reveal_type(Bar)        # type is Bar[str]
+   reveal_type(Bar[bool])  # type is Bar[bool]
+
+   class Foo(SubclassMe[int]): ...
+
+   reveal_type(Spam)  # type is <subclass of SubclassMe[int, int]>
+   Foo[str]  # Invalid: Foo cannot be further subscripted
+
+   class Baz(Generic[DefaultIntT, DefaultStrT]): ...
+
+   class Spam(Baz): ...
+   reveal_type(Spam)  # type is <subclass of Baz[int, str]>
+
+Using bound and default
+~~~~~~~~~~~~~~~~~~~~~~~
+
+If both ``bound`` and ``default`` are passed ``default`` must be a
+subtype of ``bound``.
+
+.. code:: py
+
+   TypeVar("Ok", bound=float, default=int)     # Valid
+   TypeVar("Invalid", bound=str, default=int)  # Invalid: the bound and default are incompatible
+
+Constraints
+~~~~~~~~~~~
+
+For constrained ``TypeVar``\ s, the default needs to be one of the
+constraints. It would be an error even if it is a subtype of one of the
+constraints.
+
+.. code:: py
+
+   TypeVar("Ok", float, str, default=float)     # Valid
+   TypeVar("Invalid", float, str, default=int)  # Invalid: expected one of float or str got int
+
+Function Defaults
+~~~~~~~~~~~~~~~~~
+
+The ``TypeVar``\ 's default should also be compatible with the
+parameter's runtime default if present. But they are erased to
+not have defaults when called.
+
+Implementation
+--------------
+
+At runtime, this would involve the following changes to
+``typing.TypeVar``:
+
+-  the type passed to default would be available as a ``__default__``
+   attribute.
+
+The following changes would be required to both ``GenericAlias``\ es:
+
+-  logic to determine the defaults required for a subscription.
+
+   -  potentially a way construct ``types.GenericAliases`` using a
```

Comment:
I'm not sure this is needed, and it would significantly increase the complexity of the `__class_getitem__ = GenericAlias` trick.

Reply:
It'd be a separate classmethod for GenericAlias and would only be used for a small select number of classes like Generator and maybe slice.

Reply:
`slice` is implemented in C, so it can't use this trick. I'm also not sure why Generator would need it at runtime. Can you clarify why this change is necessary?

Reply:
I was planning on having slice calling the method that I implemented to handle this case and I'd like Generator to handle this at runtime aswell as having a different behaviour in stubs is particularly annoying (I don't think I'm alone in this sentiment https://github.com/python/typing/issues/956)

Reply:
That seems like a problem independent from TypeVar defaults though. At runtime you can also write `list[1, 2]` and it will happily give you a GenericAlias. If you think that's a problem that needs to be solved, that's fine, but I don't see why this PEP is the place for it.

Reply:
I can see where you're coming from and my point kind of hinges on PEP 695 being accepted but going into the far future when typing.Generic is removed and the only way to make generics is with the new syntax I'd like to keep some level of runtime support for this. 
By default there would be little to no extra overhead (like there is now) if you didn't use defaults. However, if you used a TypeVarLike default, extra checks to just the number of arguments you passed would be performed (no type checking like there isn't now) so the correct subsitution can be performed and libraries can then inspect the generic in which ever way they see fit.

Reply:
I didn't see this comment until now sorry, anyway done.